src/storage/store/engine.rs

```diff
@@ -1707,6 +1709,86 @@ impl ShardEngine {
         }
     }
 
+    pub fn simulate_bulk_messages(
+        &mut self,
+        messages: &[proto::Message],
+    ) -> Vec<Result<(), MessageValidationError>> {
+        use std::collections::HashMap;
+
+        let mut txn_batch = RocksDbTransactionBatch::new();
+        let trie_ctx = merkle_trie::Context::new();
+        let version = self.version_for(&FarcasterTime::current());
+        let timestamp = FarcasterTime::current();
+
+        // Group messages by FID and track their original positions
+        let mut messages_by_fid: HashMap<u64, Vec<(usize, &proto::Message)>> = HashMap::new();
+        for (index, message) in messages.iter().enumerate() {
+            let fid = message.fid() as u64;
+            messages_by_fid
+                .entry(fid)
+                .or_insert_with(Vec::new)
+                .push((index, message));
+        }
+
+        // Initialize results vector with placeholder values
+        let mut results: Vec<Result<(), MessageValidationError>> = vec![Ok(()); messages.len()];
+
+        // Process each FID group
+        for (fid, fid_messages) in messages_by_fid {
```

Comment:
What's the benefit of doing this per-fid rather than doing all the messages in a single transaction?